src/irr/asset/CGLILoader.cpp

```diff
@@ -48,284 +48,385 @@ namespace irr
+				baseImageInfo.flags = static_cast<ICPUImage::E_CREATE_FLAGS>(0u);
+				baseImageInfo.samples = ICPUImage::ESCF_1_BIT;
+				baseImageInfo.extent.width = extent[0];
+				baseImageInfo.extent.height = extent[1];
+				baseImageInfo.extent.width = extent[2];
+				baseImageInfo.mipLevels = texture.levels();
+				baseImageInfo.arrayLayers = totalFaces;
+
+				auto& baseImage = core::make_smart_refctd_ptr<ICPUImage>(*(bundle.first->begin() + mipmapLevel));
+				baseImage = ICPUImage::create(std::move(baseImageInfo));
+
+				ICPUImageView::SCreationParams imageViewInfo;
+				imageViewInfo.image = baseImage;
+				imageViewInfo.format = format.first;
+				imageViewInfo.viewType = imageViewType;
+				imageViewInfo.components = format.second;
+				imageViewInfo.flags = static_cast<ICPUImageView::E_CREATE_FLAGS>(0u);
+				imageViewInfo.subresourceRange.baseArrayLayer = 0u;
+				imageViewInfo.subresourceRange.baseMipLevel = 0u;
+				imageViewInfo.subresourceRange.layerCount = totalFaces;
+				imageViewInfo.subresourceRange.levelCount = 1u;
+
+				core::vector<core::smart_refctd_ptr<ICPUBuffer>> texelBuffers;
+				auto regions = core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array<ICPUImage::SBufferCopy>>(totalFaces);
+
+				// do stuff
 
 				for (std::size_t layer = 0; layer < texture.layers(); ++layer)					// if it is an array of textures of certain type 
 					for (std::size_t face = 0; face < texture.faces(); ++face)					// if it is a cube, otherwise there is only single face
-					{	
-						// TODO, written as pseudo code
-						// core::smart_refctd_ptr<CCustomAllocatorCPUBuffer<>> buffer = CCustomAllocatorCPUBuffer<>(XSIZEX, XDATAX);	inaccessable!
-						
-						imagesOnAMipMapLevelToHandle[layer][face] = ICPUImage::create(std::move(imageInfo));
-
-						// imagesOnAMipMapLevelToHandle[layer][face]->data = texture.data(layer, face, mipmapLevel);
-						// imagesOnAMipMapLevelToHandle[layer][face]->size = texture.size(layer);
-						// rawImagesBundle = imagesOnAMipMapLevelToHandle[0][0].get();    it is wrong, but we need a way to bind rawImagesBundle and imagesOnAMipMapLevelToHandle in a way iterating through it will be valid
+					{
+						auto index = (layer * texture.layers()) + face;
+						texelBuffers.emplace_back(core::make_smart_refctd_ptr<ICPUBuffer>(baseImage->getImageDataSizeInBytes())); // I need to put data here, but
+						// core::smart_refctd_ptr<CCustomAllocatorCPUBuffer<>> buffer = CCustomAllocatorCPUBuffer<>(XSIZEX, XDATAX);	is inaccessable!
+
+						ICPUImage::SBufferCopy& region = *(regions->begin() + index);
+						region.imageSubresource.mipLevel = 0u;
+						region.imageSubresource.baseArrayLayer = 0u;
+						region.imageSubresource.layerCount = 1u;
+						region.bufferOffset = 0u;
+						region.bufferRowLength = extent[0] * texelFormatByteSize;
+						region.bufferImageHeight = 0u; //tightly packed
+						region.imageOffset = { 0u, 0u, 0u };
+						region.imageExtent = {extent[0], extent[1], extent[2]};
```

Comment:
you need to vary `mipLevel` and `bufferOffset` accordingly

Also are you sure GLI will give tightly packed image rows?

Reply:
Finally `extent` shrinks for each mip level, the ICPUImage or IImage has a utlity function that actually tells you the extent of a given mip-level.

Reply:
> the ICPUImage or IImage has a utlity function that actually tells you the extent of a given mip-level.

well, I get extent relative to mipmapLevel by loaded GLI texture that has a capability of getting an extent for each mip level, but I will remember it

Reply:
https://github.com/buildaworldnet/IrrlichtBAW/pull/402/files#diff-28a7c4ae6aba8ec71c707eede79b01e9R206 if a swizzle is blue, it will be assigned to red so forth, so it seems to me it is correct

Reply:
what is this last comment about!

Reply:
status? @Przemog1